### ISSUE TITLE
res.sendFile doesn't send static files

### ISSUE THREAD

**Issue by jdrydn** [serverless/cloud#159]
**tl;dr:** Trying to get a SPA running from a subdirectory (`/app`), thought I could use `res.sendFile` to serve the `./static/app/index.html` for `GET /app/*` but it looks like the dev instance has a different set of `static/` files to the personal instance.

----

**Use-case:** Serving a SPA from a subdirectory (`/app`) on Serverless Cloud:

- Requests to `/app` render the SPA fine, as [per the docs](https://www.serverless.com/cloud/docs/apps/static-assets#static-html-pages) `/app -> /app/index.html` :+1:
- Clicking to a new page (e.g. `/app/about`) is also fine, as this click was intercepted client-side & the relevant page shown by the SPA :+1:
- Refreshing the page at this point creates a request to the Serverless Cloud function for `/app/about`, where there isn't a relevant `/app/about.html` or `./app/about/index.html`, and returns a 404 Not Found :+1:

Under [Serving Serverless Storage files](https://www.serverless.com/cloud/docs/apps/api#serving-serverless-storage-files) in the docs (emphasis mine):

> To send either locally stored files or files stored via Storage, use `res.sendFile`. **`sendFile` will first check the local directory - if the supplied path exists here, the local file will be sent**. Otherwise, `sendFile` will redirect to a download url for a given path, if it exists, from Storage.

Now, whilst "local directory" isn't directly clear, I assumed that local directory would include items in `./static`. And within my personal instance, the following code works fine:

```js
import path from 'path';
import { api } from '@serverless/cloud';

api.use('/app', (req, res) => {
  res.sendFile(path.resolve(__dirname, './static/app/index.html'));
});
```

However, when running `cloud deploy dev` to deploy this to a dev instance, the above code always returns `404 Not Found`. After a little debugging with `console.log` I can confirm the route is reached, the error appears to be within `sendFile`.

Bringing `fs` into the mix to inspect the situation further, I can confirm that:

- `process.env.PWD === __dirname` are the same value (`/tmp/task`) on both personal & dev instances.
- Checking with `fs.readdirSync`,  the `./static` folder is present within the deployed function code on both personal & dev instances.
- On the personal instance, `./static/app/index.html` is present. But on the dev instance, `./static/app/index.html` is missing? In fact, **all the files** within `./static` are missing, just the folder structure remains.

To visualise this, the following code produces the following output on the personal & dev instances:

```js
api.use('/app', async (req, res) => {
  console.log('APP ROUTE', req.url);
  console.log(__dirname, process.env.PWD);
  console.log([
    fs.readdirSync(path.resolve(__dirname, './static')),
    fs.readdirSync(path.resolve(__dirname, './static/app')),
    fs.readdirSync(path.resolve(__dirname, './static/app/img')),
  ]);

  res.status(200).send('HELLO WORLD');
});
```
```
# PERSONAL INSTANCE
06:15:08 ─ APP ROUTE /about
06:15:08 ─ /tmp/task /tmp/task
06:15:08 ─ [ [ 'app' ], [ 'css', 'img', 'index.html', 'js' ], [ 'logo.82b9c7a5.png' ] ]

# DEV INSTANCE
06:16:22 ─ APP ROUTE /about
06:16:22 ─ /tmp/task /tmp/task
06:16:22 ─ [ [ 'app' ], [ 'css', 'img', 'js' ], [] ]
```

The issue appears to be within deploy process, removing (the contents of) `./static` to make for a smaller package size.

Should note though, a suitable workaround involves:

```js
import axios from 'axios';
import { api } from '@serverless/cloud';

// A web request per function start - not ideal
const appPage = axios.get('/app/index.html', {
  baseURL: process.env.SERVERLESS_CLOUD_URL,
  responseType: 'text',
  validateStatus: status => status === 200,
});
api.use('/app', async (req, res) => {
  const { data } = await appPage;
  res.status(200).set('Content-Type', 'text/html').send(data);
});
```

---

In terms of next steps / questions:

- Could a member of the Serverless Cloud team define "local directory"?
- Could/should the Serverless `storage` class return static files as buffers?
- Should the `./static` folder even be present?
- I'd offer to make a PR with `sendFile` support for static files, or `storage` class support for static files, but the source is obviously closed.

Apart from this strange bug/issue, I'm very much enjoying the experience of Serverless Cloud whilst building my prototype project! :heart:

**Comment by rschick** [comment]
Thanks for bringing this to our attention. It looks like there could be an issue with sendFile and static files. You are correct that in stage deployments static files are not in the local FS, but sendFile should handle this.

Have you tried `http.on(404)`?

```
import { http } from "@serverless/cloud";
 
http.on(404, "app/index.html");
```

**Comment by jdrydn** [comment]
Hey @rschick, thanks for getting back to me.

> You are correct that in stage deployments static files are not in the local FS, but sendFile should handle this.

Okay cool - so that would suggest there's a bug in `sendFile` where it doesn't properly send static files (which I imagine are just ~on S3~ in storage?)

> Have you tried `http.on(404)`?

This is a perfect solution for hosting a SPA in front of an API-like project, but I'm afraid that's not a solution for this project, as the SPA is (currently) a sub-directory & the rest of the routes are to be handled by the server. Roughly, this is what I was aiming for:

```
/app
  - SPA under ./app
  - Built & deployed from ./static/app
/api
  - API from ./api/router.ts
/
  - Main site rendered from ./site/router.ts
  - Could be Liquid, Handlebars or another server-side templating language.
  - Or could be rendered with Jekyll etc. before deploy time.
```

**Comment by jdrydn** [comment]
> ~I'm afraid that's not a solution for this project~

That's not entirely fair - I'm now looking to see if I can expand the SPA to include some SSR packages & rework the client-side store & auth to cover a server environment, if so then running the SPA as the whole frontend & `http.on(404, ...)` would work fine.

**Comment by rschick** [comment]
@jdrydn Ok I see what you mean now. I'll get the team to look into this.
